File: Array/LeetCode142.go
package array

type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	slow, fast := head, head
	for ok := true; ok; ok = (slow != fast) {
		if fast == nil || fast.Next == nil {
			return nil
		}
		fast = fast.Next.Next
		slow = slow.Next
	}
	fast = head
	for fast != slow {
		slow = slow.Next
		fast = fast.Next
	}
	return fast
}

File: Array/LeetCode142.cc
#include<iostream>
#include<vector>
using namespace std;


struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *slow = head;
        ListNode *fast = head;

        do{
            if(!fast || !fast->next){
                return nullptr;
            }
            
            fast = fast->next->next;
            slow = slow->next;

        }while(fast != slow);   

        fast = head;

        while(fast != slow){
            slow = slow->next;
            fast = fast->next;
        }
        return fast;
    }
};